packages/palette/src/elements/Shelf/Shelf.tsx

```diff
@@ -0,0 +1,289 @@
+import { themeGet } from "@styled-system/theme-get"
+import React, {
+  Children,
+  createRef,
+  isValidElement,
+  useEffect,
+  useMemo,
+  useRef,
+  useState,
+} from "react"
+import styled from "styled-components"
+import { useCursor } from "use-cursor"
+import { visuallyDisableScrollbar } from "../../helpers/visuallyDisableScrollbar"
+import { ChevronIcon } from "../../svgs"
+import { Box, BoxProps } from "../Box"
+import { CELL_GAP_PADDING_AMOUNT, paginateCarousel } from "../Carousel"
+import { CarouselBar } from "../CarouselBar"
+import { Clickable } from "../Clickable"
+import { FullBleed } from "../FullBleed"
+
+/** ShelfProps */
+export type ShelfProps = BoxProps & {
+  snap?: "none" | "start" | "end" | "center"
+  children: JSX.Element | JSX.Element[]
+  onChange?(index: number): void
+  onProgress?(progress: number): void
+}
+
+/**
+ * A Shelf is a new kind of carousel...
+ */
+export const Shelf: React.FC<ShelfProps> = ({
+  children,
+  snap = "none",
+  onChange,
+  onProgress,
+  ...rest
+}) => {
+  const cells = useMemo(
+    () =>
+      Children.toArray(children)
+        .filter(isValidElement)
+        .map((child) => ({ child, ref: createRef<HTMLLIElement>() })),
+    [children]
+  )
+
+  const containerRef = useRef<HTMLDivElement | null>(null)
+  const viewportRef = useRef<HTMLDivElement | null>(null)
+
+  const [mounted, setMounted] = useState(false)
+  const [pages, setPages] = useState([0])
+  const [progress, setProgress] = useState(0)
+  const [offset, setOffset] = useState(0)
+
+  const init = () => {
+    if (containerRef.current === null || viewportRef.current === null) return
+
+    // Set page-stops
+    const { current: viewport } = viewportRef
+    const values = cells.map(({ ref }) => ref.current.clientWidth)
+    setPages(paginateCarousel({ viewport: viewport.clientWidth, values }))
+
+    // Set offset to accomodate full-bleed and line up initially with page-margins
+    const { current: container } = containerRef
+    const { x } = container.getBoundingClientRect()
+    setOffset(x)
+    setMounted(true)
+  }
+
+  useEffect(() => {
+    init()
+
+    window.addEventListener("resize", init)
+    return () => {
+      window.removeEventListener("resize", init)
+    }
+  }, [cells])
+
+  const { index: pageIndex, setCursor } = useCursor({
+    max: pages.length,
+  })
+
+  // Keep page index in sync with scroll position
+  useEffect(() => {
+    if (viewportRef.current === null) return
+
+    const { current: viewport } = viewportRef
+
+    const handler = () => {
+      const nearestPage = pages.find((currentPage, i) => {
+        const nextPage = pages[i + 1] ?? Infinity
+        return (
+          viewport.scrollLeft >= currentPage && viewport.scrollLeft < nextPage
+        )
+      })
+
+      setCursor(pages.indexOf(nearestPage))
+
+      const __progress__ = Math.ceil(
+        (viewport.scrollLeft * 100) / (pages[pages.length - 1] - 1)
+      )
+
+      setProgress(__progress__)
```

Comment:
Interesting, so whats happening here is yr setting a computed value which then outputs `progress` which is then passed up through the system. Took a sec to understand due to the shadowed var. Maybe naming it something like `computedProgress` could reduce some friction here?